[hue] Fix multiple state updates (API v2)

jlaur opened the pull request:
The overall goal of this pull request is to address issues with unneeded state updates that can happen under certain circumstances.

This can cause:
- Flickering in UI's.
- Misbehavior from rules (see chapter **Context** in #15700).
- A bit of a mess in persisted values.

Three root causes have been identified:
- Multiple messages wrapped in one event are processed sequentially (one by one), causing intermediate state updates.
- _State sent as multiple events where dimming information is missing in first event._ This has been extracted to #15995.
- _Auto update policy is **default**, so commands will trigger state update immediately rather than awaiting the final state from the bridge._ This fix for this has been extracted to #15984.

Given an initial brightness of 0 and this command:
```shell
openhab> openhab:send Brightness 100
Command has been sent successfully.
```

Yielding this payload:
```json
[
	{
		"creationtime": "2023-10-05T19:32:40Z",
		"data": [
			{
				"id": "00000000-0000-0000-0000-000000000001",
				"id_v1": "/lights/1",
				"on": {
					"on": true
				},
				"owner": {
					"rid": "00000000-0000-0000-0000-000000000000",
					"rtype": "device"
				},
				"type": "light"
			},
			{
				"dimming": {
					"brightness": 100.0
				},
				"id": "00000000-0000-0000-0000-000000000001",
				"id_v1": "/lights/1",
				"owner": {
					"rid": "00000000-0000-0000-0000-000000000000",
					"rtype": "device"
				},
				"type": "light"
			}
		],
		"id": "00000000-0000-0000-0000-000000000002",
		"type": "update"
	}
]
```

will result in the following events:
```
2023-11-26 00:14:12.970 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'Brightness' received command 100
2023-11-26 00:14:12.971 [INFO ] [penhab.event.ItemStatePredictedEvent] - Item 'Brightness' predicted to become 100
2023-11-26 00:14:12.972 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Brightness' changed from 0 to 100
2023-11-26 00:14:13.085 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Brightness' changed from 100 to 0
2023-11-26 00:14:13.086 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Brightness' changed from 0 to 100
2023-11-26 00:14:13.086 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'Switch' changed from OFF to ON
```

The reason for the state update back to 0 is that this part of the message is fully processed with channel update:

```json
				"on": {
					"on": true
				},
```

before processing the next message:
```json
				"dimming": {
					"brightness": 100.0
				},
```

This means that the already cached brightness will be applied rather than the brightness received in the separate message. The final state will be correct, but an additional update is made.

## Testing

In addition to the tests documented above, some manual regression tests performed:
- Motion sensor channels are working.
- Brightness can be changed for Thing with Zigbee connectivity issue.
- Thing is updated according to Zigbee connectivity status.

Fixes #15700
Related to #15983

andrewfg commented:
@jlaur I only just noticed this PR today. At first glance I would say that this is NOT a solution. Because I think there is actually NOT a problem that could or should be 'solved' within OH. As mentioned elsewhere the Hue system handles different DTOs orthogonally, and IMHO the binding must also handle them orthogonally. And attempting to munge one more more orthogonal DTOs into a common state change is not correct.

andrewfg commented:
^
Specifically I can think of several hard examples where DTO munging would be most dangerous..
- Push buttons risk to lose the individual button down, button hold, button up events
- Security and tamper contacts risk to lose intermediate alarm states
- Motion detectors ditto.
- Rotary dials (where each event represents number of clicks offset from the prior event) risk to lose intermediary clicks.

Also as said before there are three ways that the bridge can send multiple events .. namely
1. Combine two DTOs in one event
2. Pack two DTOs into two sequential packets in the same event stream
3. Send two DTOs one after another in two sequential event streams

Your 'solution' only tries to address 2. above. 
Push buttons, security devices, and rotary dials use sometimes 2. sometimes 3. and sometimes even both..

andrewfg commented:
> in some sort of pre-processor logic.
> it should only apply to color, brightness and on/off.

Ok. That makes sense. But I would ask that the `abcOnly` channels should retain the original (non pre-processed) logic.

Maybe just to point out one thing (in case you overlooked it): in your examples above you imply that the 'cache' is the channel HSBType value, which is not strictly true .. the cache actually comprises a XY pair, a Dimming, and an on-off attribute .. and as you imply (particularly when 'B' is close to zero, or when 'XY' is close to the edge of the Gamut) there will be round trip conversion issues between 'XY' and 'HS' .. and perhaps also between 'D' and on-off (particularly when 'D' is close to the minimum dimming level) .. so I advise that the pre-processing should cache and merge the 'XYDO' values (rather than the HSBType value) i.e. you should only make a one way final conversion 'XYDO' => HSB when setting the HSBType on the channel.

jlaur commented:
@andrewfg - are you still providing backports to 3.4.x?

andrewfg commented:
> are you still providing backports to 3.4.x?

No.

jlaur commented:
> * what other PR are you waiting for?

See last line of description. I found a bug causing auto update policy to be ignored when defined at channel level. This directly impacts this PR, unfortunately.

> * apart from that, is this PR now ready for me to review?

Almost, but it seems there is some GitHub hiccup at the moment:

![image](https://github.com/openhab/openhab-addons/assets/19519842/7a402850-959e-40fb-a870-e7152d29314a)

Everything is in the branch though, so you can already check everything in the last commit. The only change missing in this PR from the pushed branch is some auto update policies for additional channels.

> * it seems to be a lot bigger, than the original scope of the PR => do you want to change the title?

The scope should be what the title currently says. The only thing is that the issue turned out to have two different root causes. I tried to explain the thoroughly in the description, but please let me know if you need anything else. The linked issue only mentions one of those root causes, but they both have similar consequences. Therefore I decided to extend the original scope.

It was a long way for me to get it to the current state, with a lot of testing and head banging, and also with a lot of changes now reverted. One thing that took me a while to understand is that all thing handlers receive all resources and maintain their own caches. Each has its own `Map`, but the contained `Resource` objects are actually shared. This lead to some problems until I realized that the pre-processing had to be done by the bridge before iterating through the handlers and sharing the modified list of resources. I was finally able to get everything working yesterday and reverted a lot of copy constructors etc. that are now outside the scope of this PR.

andrewfg commented:
> all thing handlers receive all resources and maintain their own caches. Each has its own Map, but the contained Resource objects are actually shared.

Only partly true. Each thing has its own unique resource which contains the full state of the thing, and partial data from the incoming (shared) resources is copied from them into the own thing specific unique resource.

jlaur commented:
> Only partly true. Each thing has its own unique resource which contains the full state of the thing, and partial data from the incoming (shared) resources is copied from them into the own thing specific unique resource.

What I mean is that the references to the individual DTO's are copied (by `Setters.setResource`). So those objects will actually be shared between the different thing handlers, since they are not copied by value. The `Resource` object itself is also stored in `serviceContributorsCache`, i.e. the same `Resource` object that is part of `List<Resource>` that all handlers will receive. Or am I missing something?

andrewfg commented:
@jlaur thinking about this further .. this PR is really far too complicated with too many changes risking to break existing functionality. I am pretty sure that I can achieve what you want in about one tenth of the code changes. I can visualize the correct solution in my head already, but I think it is probably easier for me to write the code myself rather than try to explain it via a review process. So I think I will create a new branch myself and let you have a look at it. It may take me a couple of days, so please be patient.

andrewfg commented:
@jlaur here is the my version of the PR https://github.com/openhab/openhab-addons/pull/15969/files -- note not yet tested.

andrewfg commented:
@jlaur please look at my alternative PR; I am really sure that you are on the wrong track.

jlaur commented:
> @jlaur please look at my alternative PR; I am really sure that you are on the wrong track.

Please be more concrete. And please also read the description and implementation I have provided here. I'm open to improve/replace the algorithm provided here:

https://github.com/jlaur/openhab-addons/blob/33d8a7d8f7980bfefebe777e3ac6617ef42df10b/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java#L354-L389

which I guess is essentially what you would like to change?

I will await your tests before investing in a deeper look and testing, since I already spent quite some time testing my version (which I also have running in my production system).

When I'm convinced your version is superior, I could copy it in parts or in full into this PR and verify it with the unit test coverage I have provided.

andrewfg commented:
> Please be more concrete

1. The need for merging only applies to SSE event partial resources (full data resources already have full data obviously)
2. The need for merging only applies to the case when there are two Resources, both of type LIGHT, and both with identical resource Id, and both within the same SSE packet
3. The merging must be done in the bridge and not in the thing; it applies to all things that may use the merged event resource
4. It is not necessary to do deep copies during the merge
5. It is not necessary to change the device thing handler code
6. Setters.setResource is therefore fully sufficient for the merge
7. My code is about 35 lines in one single java class; yours requires far more code, in far more java classes.
8. Your code is changing the sequencing, so risks to break something else. Mine does not do that.

andrewfg reviewed (changes requested):
This PR is not the right approach. There is a better solution here https://github.com/openhab/openhab-addons/pull/15969/files

jlaur commented:
> 1. The need for merging only applies to SSE event partial resources (full data resources already have full data obviously)

Agreed, and this is what I have implemented.

> 2. The need for merging only applies to the case when there are two Resources, both of type LIGHT, and both with identical resource Id, and both within the same SSE packet

Almost agreed (and implemented) - however, doesn't it also apply to **grouped_light**?

> 3. The merging must be done in the bridge and not in the thing; it applies to all things that may use the merged event resource

Agreed, and this is what I have implemented.

> 4. It is not necessary to do deep copies during the merge

Agreed, and this is what I have implemented. The only exception in my implementation is the Zigbee connectivity status which I need to copy because I'm creating a new object.

> 5. It is not necessary to change the device thing handler code

Agreed. I changed it initially because I **was** in the wrong track until a few days ago. However, I left the refactoring that improved the code quality by eliminating code duplication.

> 6. Setters.setResource is therefore fully sufficient for the merge
> 7. My code is about 35 lines in one single java class; yours requires far more code, in far more java classes.

Well, I wouldn't agree to measure only by number of code lines. I would opt for the most readable and robust implementation. If we are to compare, and just to avoid any misunderstandings, my algorithm is also 35 lines (in Setters) + 37 lines for supporting methods in `Resource`. That includes empty lines and Javadoc.

> 8. Your code is changing the sequencing, so risks to break something else. Mine does not do that.

It will move HSB-related DTO's to a `Resource` at the end, but that doesn't matter as there is no dependency on the order.

To conclude, and sorry for being direct, but it seems like you might have been following some previously drafted code and actually not having read the final implementation? As said, I'm open for discussing the actual merge implementation, and when you are finished with your version, I can also test it.

Surprisingly, you have not mentioned the auto update policy issue at all, which is half of the fix.

> This PR is not the right approach. There is a better solution here https://github.com/openhab/openhab-addons/pull/15969/files

Tbh, I think you are being quite disrespectful here. I have provided a working fix, even with quite detailed explanations and test documentation. You have not bothered to review it or even fully read/understand it, because your "gut feeling" is that the approach is wrong. Instead you dismissed all of it, created a different version with possibly half the fix, no test coverage, and just claim that is better without any reasonable kind of explanation.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:
```diff
@@ -528,13 +529,15 @@ public void onResourcesEvent(List<Resource> resources) {
-                resources.forEach(resource -> {
-                    ((Clip2ThingHandler) handler).onResource(resource);
-                });
+            if (thing.getHandler() instanceof Clip2ThingHandler clip2ThingHandler) {
```
Do you want to do all these conversions from Java 9 to Java 17 in a separate PR? (There are probably more than just this one..)

jlaur replied:
I only changed it here because I was in the neighborhood. If there will be another PR for redoing those improvements, I will simply resolve this conflict.

I will not do general Java 17 improvements in context of this PR. i.e. those previously done in #15520 and [rejected by you](https://github.com/openhab/openhab-addons/pull/15520#issuecomment-1698194898).

andrewfg replied:
I don't mind if this change remains in the PR. Personally I have no plans to make another PR to convert Java 9 code styles to Java 17 styles, since Java 17 did not deprecate Java 9 styles, and so there is no need to change anything.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -52,6 +57,8 @@
 @NonNullByDefault
 public class Setters {
 
+    private static final Set<ResourceType> LIGHT_TYPES = Set.of(ResourceType.LIGHT, ResourceType.GROUPED_LIGHT);
```
See my other discussion about grouped_light -- is it proven that grouped_light has a problem? Why would you need to fix something if it is not proven to be broken? The two API's are different, as shown below, so there are at least some grounds to suspect that the bridge would handle them different. Specifically grouped_light  does not use ColorXy or ColorTemperature..

- https://developers.meethue.com/develop/hue-api-v2/api-reference/#resource_light_get
- https://developers.meethue.com/develop/hue-api-v2/api-reference/#resource_grouped_light_get

jlaur replied:
The issue with the **light** resource was also not "proven" by you - I'm the only one of us having experienced the issue, which is specifically caused by having **on** in one resource and **dimming** in another. Both are also part of **grouped_light**. So I see no reason **not to fix** this.

I could probably reproduce it by creating a zone for my dimmer having the issue already, but in any case we don't want this coming back to us in a future bug, even if it currently seems theoretical with the current firmware and the particular devices and configurations we are using for testing. Especially because it required no additional effort or complexity to fix this as well.

jlaur replied:
The other way around.

If **grouped_light** has the same issue, which logically would make sense since it can also contain **on**/**dimming** values, we should rather include the fix for this scenario as well.

On the other hand, if **grouped_light** doesn't have the same issue, the fix will do no harm. In fact, it won't even have any impact, since nothing would be merged.

I really don't see how this could create a bug that wouldn't also be created for **light** then.

andrewfg replied:
Ok. Closing this issue.

jlaur commented:
> I think it would be pointless to do a formal review of your code, since it would only be fine tuning what I consider to be fundamentally wrong.

I agree that was pointless. Therefore I'll reply only to the comments with substance, and resolve the ones providing no value but stating "this is different than my implementation".

jlaur commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:
```diff
@@ -528,13 +529,15 @@ public void onResourcesEvent(List<Resource> resources) {
 
     private void onResourcesEventTask(List<Resource> resources) {
-        logger.debug("onResourcesEventTask() resource count {}", resources.size());
+        int numberOfResources = resources.size();
```
I have updated my implementation based on inputs from:
- My analysis of your suggested method above.
- Your comments elsewhere.
- The Hue API documentation.

In short, I agree that separating HSB-containing resources into new resources is not needed, hence the cloning has been removed. I have also made some other simplifications and I think the readability is improved.

> The only exception to the above would be if the Resource also contains a sub-DTO for other than pure light variable state DTOs e.g Alerts, or Effects. In which case we do not merge the two resources, but rather let them both be processed separately.

I'm handling this by still merging HSB-related fields, but leaving the non-HSB fields separate. See test case `mergeLightResourcesMergeHSBFieldsDoNotRemoveResourceWithEffect`.

andrewfg replied:
> agree that separating HSB-containing resources into new resources is not needed, hence the cloning has been removed

Yay!

> merging HSB-related fields, but leaving the non-HSB fields separate

See my above comments concerning a|) ColorTemperature, and b) 'complex' resources. In short my believe is that if both resources contain "pure light state" 'hsbFields' fields ONLY, then they can safely be merged; whereas if they contain a complex mix of 'hsbFields' and 'nonHsbFields' we should rather not tamper with them.

jlaur commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +348,43 @@ public static Resource setResource(Resource target, Resource source) {
+     * Extract on/dimming/color fields from light and grouped light resources
+     * and merge them into separate resources.
+     */
+    public static Collection<Resource> mergeLightResources(Collection<Resource> resources) {
```
`Setters.setResource` will not overwrite existing fields, so in principle, if two subsequent dimming values would arrive in sequence, the last one would not be taken. It doesn't matter much though since there are only three relevant fields to merge: on, dimming and color, so this logic can easily be handled without all the other merging done by `setResource`.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +349,67 @@ public static Resource setResource(Resource target, Resource source) {
+                resource.setColorXy(null);
+            }
+
+            if (!resource.hasItems()) {
```
I am not sure if I agree with your approach here.

Your approach (finally) is to de-link the 'hsbFields' references from resources later in the list and re-link them to the first resource in the list. (Which is actually what I was proposing). This is good. But this means that any references to 'nonHsbFields' are left behind ('orphaned') in the later resource. This is where I have a doubt.

My feeling is that if the resources have ONLY 'hsbFields' and NO 'nonHsbFields' then one can indeed safely de-/re-link the 'hsbFields' references, and remove the now empty resource from the list. HOWEVER if the resources have BOTH 'hsbFields' and ALSO 'nonHsbFields' (a 'complex' resource), then it is better NOT to de-/re-link any of the field references; and rather let both resources pass through verbatim in this list. My logic is that such a 'complex' resource with mixed fields may contain perhaps more complex logic, and I think we should rather not mess around with such complex objects.

jlaur replied:
The only reason for merging anything rather than keeping it separate and processing it sequentially (like before this PR) is because we have a dependency between **on**, **dimming** and **color** on the openHAB side because of the combined HSB.

Therefore the objective is to mitigate multiple state updates by merging such resources. For anything else, there is strictly no need to merge, so therefore I kept the logic strictly to fulfil this requirement.

I also checked `updateChannels`, and there isn't any problems with this, it simply processes the objects that exists. This also makes perfect sense, because we receive sparse resources and should be able to handle them individually already.

andrewfg replied:
> we have a dependency between on, dimming and color .. Therefore the objective is to mitigate multiple state updates by merging such resources.
> For anything else, there is strictly no need to merge, so therefore I kept the logic strictly to fulfil this requirement.

Indeed. My point relates to your "anything else": If either of the resources has on, dimming and color and ALSO **_anything else_** then I think there may be a risk (now or future) if we mess around with the resources. So my suggestion is as follows..
- Both resources have on, dimming and color ONLY => merge Ok
- Either resource has on, dimming and color and ALSO **_anything else_** => do not merge

jlaur replied:
> To be more precise, I fear that the 'merge some fields but not others' (half pregnant) approach may be risky..

Why do you think that? There should be no dependencies between the objects, which we already process individually as we receive them. The only dependency is the one we have because of HSB. There shouldn't be any difference between us separating them apart and receiving them individually straight from the source, i.e. it would be quite unexpected that this would cause any problems. Can you please give an example?

> Either resource has on, dimming and color and ALSO anything else => do not merge

That would break the fix. Simplest example:
- Resource A: on=on
- Resource B: dimming=75, effects=something

I don't know if this can happen, but I wouldn't make any assumptions that it can't, and also that's not the point. With current logic we'll process:
- Resource A: on=on, dimming=75
- Resource B: effects=something

This will work. If we wouldn't merge in this case, we would still have the bug under those circumstances.

andrewfg replied:
> I don't know if this can happen, but I wouldn't make any assumptions that it can't

Me neither. That IS my point ..

> This will work. 

Are you not being over-confident? The [API](https://developers.meethue.com/develop/hue-api-v2/api-reference/#resource_light_get) for LIGHT includes many elements, and I am simply wondering if either now or in future the bridge may pack combined 'complex' elements into a single specific resource so that they may be read in combination (I am thinking of say dynamics, alert, mode, signalling or gradient (some of which we don't currently use) which might need to be interpreted in combination with a colorxy, dimming, or onoff element; and if we had 'stolen' that element from resource #2 to put in resource #1, then the resource #2 resource might lose its meaning.

I am just worried that by doing something non transparent here and now, we might be creating potentially a quite hard to debug problem later.

However to take the other side of the argument ;) -- I have always argued that Hue DTOs are orthogonal, so the other side of the argument would be to not expect any such non orthogonality and not be upset if any non orthogonality would be broken by this fix.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +349,67 @@ public static Resource setResource(Resource target, Resource source) {
+                indexedResource.setColorXy(colorXy);
+                resource.setColorXy(null);
+            }
+
```
I think you should probably do the same as above (for OnOff, Dimming, ColorXy fields) for the ColorTemperature field. Admittedly it is not directly HSB related, but if the light has been commanded from a colorXy co-ordinate that is off the 'white color locus' to one that is on the locus (or vice versa) then the SSE DTO will contain both a ColorXy field and also a ColorTemperature field. And if one de-/re-links the former field, then one should also de-/re-link the latter.

Note: this is slightly linked to my comment below about merging / not merging of 'complex' resources.

jlaur replied:
This is related to the discussion above, so to not duplicate I'll resolve this one.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/Resource.java:
```diff
@@ -121,6 +121,31 @@ public Resource(@Nullable ResourceType resourceType) {
+     *
+     * @return true is resource doesn't contain any state information
+     */
+    public boolean hasItems() {
```
As mentioned before, please rename method as 'hasNonHSBField'

jlaur replied:
Resolving this comment to not have duplicate discussions, I have responded in the other one.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/Resource.java:
```diff
@@ -121,6 +121,31 @@ public Resource(@Nullable ResourceType resourceType) {
+     *
+     * @return true if resource has any HSB component
+     */
+    public boolean hasHSBComponent() {
```
As mentioned before, please rename method as 'hasHSBField'

jlaur commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +349,67 @@ public static Resource setResource(Resource target, Resource source) {
+                continue;
+            }
+
+            // Full state takes precedence, invalidate previous sparse resources.
```
Ah, thanks, that simplifies it further. I had missed:
https://github.com/openhab/openhab-addons/blob/4fbb6c965566ae15ee24822882fc6e35ca30bbfb/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java#L929-L930

jlaur commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +349,67 @@ public static Resource setResource(Resource target, Resource source) {
+        }
+
+        // "Garbage collect" resources with no items left.
+        resources.removeIf(r -> LIGHT_TYPES.contains(r.getType()) && !r.hasItems());
```
This is not what the method checks, it actually checks if the resource has any items of its corresponding resource type. The word "item" was taken from the the API documentation and was used deliberately as "logic" term rather than "Java" term.

andrewfg replied:
The term 'items' can be confused with OH items. We are referring to sub-DTO fields in the main DTO. So let's make rhat clear. You already accepted my proposal to change hasHSBField, so the naming should be consistent across those methods. If the hasXYXField includes more scope than hasNonHSBField would do, then rename it to whatever 'hasXYZField' most aptly describes it. 'hasItems' does not describe enough..

However depending on the other discussion, you may need to replace that method anyway with something like 'isComplexLightResource'..

jlaur replied:
Adopted the term "field".

andrewfg replied:
PS probably even clearer would be..

- hasHSBJsonObject() resp. hasLightJsonObject()

jlaur commented:
Since I am not able to progress with openhab/openhab-core#3888, I have extracted the auto update policy part into #15984.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +349,60 @@ public static Resource setResource(Resource target, Resource source) {
+            }
+
+            if (!resource.hasAnyRelevantField()) {
+                iterator.remove();
```
@jlaur I think you are trying to do the “garbage collection” twice — once by iterator.remove inside the loop, and again via removeIf outside the loop.

.. and if you decide to do it via removeIf, then it may be that an iterator is not required, and the loop could be executed more simply via a forEach syntax.

jlaur replied:
Correct. I went with the additional `iterator.remove()` for performance reasons, because in situations where it's possible to decide to remove the object straight away, this will avoid an additional traversal of the collection.

The performance gain might be negligible, so if you feel strongly for removing this optimization, I'm fine with that.

andrewfg replied:
I have no strong feelings about the optimisation. But if the remove is done within the loop then the external removeIf is dead code so it can be deleted.

andrewfg replied:
I suggest to delete the removeIf line. And as you say we can modify hasAnyRelevantField in future as required.

.. maybe hasAnyRelevantField should be called hasAnyLightRelevantField ??

jlaur replied:
I have removed the removal of irrelevant resources.

> .. maybe hasAnyRelevantField should be called hasAnyLightRelevantField ??

I think it's fine as is. Currently it supports only **light** and **grouped_light**, but it can easily be extended to support other types when needed.

andrewfg replied:
The current method name is not really self documenting (relevant to what) .. a fully self documenting method name would be hasFieldsNotRelevantToMergingofPureLightStateInLightOrGroupedLightResources 

.. perhaps you can explain it in the Java doc?

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/helper/Setters.java:
```diff
@@ -341,7 +349,60 @@ public static Resource setResource(Resource target, Resource source) {
+            }
+
+            if (!LIGHT_TYPES.contains(resource.getType()) || !resource.hasHSBField()
+                    || resource.equals(indexedResource)) {
```
Also I think you can remove the resource.equals() check (which is quite a dangerous check if someone would override the equals method in future), and instead insert a continue after line 374 ??

andrewfg replied:
Oh, I see you did that already..

jlaur replied:
I agree it's dangerous and I didn't like it one bit. Initially I did `resource == indexResource`, but that triggered a SAT warning. When I tried to suppress that, I instead got a compiler warning about the suppresion. Since we didn't have an override, I made the dubious choice to use `equals`. I'm glad we can now remove it after the other simplifications, but I still would like a solution for this generic problem with SAT for future cases.

andrewfg commented on bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/api/dto/clip2/Resource.java:
```diff
@@ -202,9 +205,19 @@ public State getBrightnessState() {
-                double brightness = Objects.nonNull(on) && !on.isOn() ? 0f
-                        : Math.max(0f, Math.min(100f, dimming.getBrightness()));
+                double brightness = Math.max(0f, Math.min(100f, dimming.getBrightness()));
+                if (on != null) {
```
@jlaur I think I may have misunderstood what you said in the Issue, so let me just check it here..

When the binding starts, it always does an initial download of the full state of all resources. And then the state is updated by delta via individual SSE messages.

Therefore I would assume that even if a light is initially off, the initial full state download would include the last known Dimming state of the lamp. i.e. the full state plus delta should never actually have the case where Dimming == null. => Are you saying that you have actually seen such cases of full plus delta Dimming == null in real life?

jlaur replied:
> Are you saying that you have actually seen such cases of full plus delta Dimming == null in real life?

Actually, there is no change regarding this, and no, I haven't seen that.

But I have seen this in an initial state response just after turning the dimmer on at low dimming by physical button:

`GET https://hue.local/clip/v2/resource/light/00000000-0000-0000-0000-000000000001`

```json
            "on": {
                "on": true
            },
            "dimming": {
                "brightness": 0.0
            },
```

Interesting for this case is that it is shown as 1% brightness in the Hue app. With the fix provided here, we will now interpret as 100% (rather than off/0% as before). This works better for the case I originally detected where the dimmer is turned fully on and the events arrive separately as "on" and "100%", because it will then only cause one state update. But as we discussed in the issue, there is no really good choice anyway, because we simply don't know the actual value.

> .. and what about the case of On/Off only lamps?

They don't have a brightness channel, and the code in unchanged, so it should work fine?

jlaur replied:
These two testcases are failing without the fix, all the others pass in both versions:
- getBrightnessStateWhenOnAndDimming0ReturnBrightness100
- getBrightnessStateWhenOnAndDimmingMinus1ReturnBrightness100

andrewfg replied:
>                "brightness": 0.0
> that it is shown as 1% brightness in the Hue app
> With the fix provided here, we will now interpret as 100%
> we simply don't know the actual value

Hmm. I think, therefore, that we should interpret it exactly as the Hue App interprets it. Namely On @ Dimming=0% should be interpreted as On @ Dimming=minimum dimming level.

> it will then only cause one state update

And .. the merging of consecutive SSE events (in this PR) should mean that the first On @ Dimming=minimum dimming level in resource #1 would then be immediately resolved to On @ Dimming=whatever SSE value from resource #2. Or??

jlaur replied:
>> Hmm. I think, therefore, that we should interpret it exactly as the Hue App interprets it. Namely On @ Dimming=0% should be interpreted as On @ Dimming=minimum dimming level.

> There is no minimum dimming level provided by this device (at least which is compatible with Hue), so it's quite a corner case.

>> it will then only cause one state update

> And .. the merging of consecutive SSE events (in this PR) should mean that the first On @ Dimming=minimum dimming level in resource 1 would then be immediately resolved to On @ Dimming=whatever SSE value from resource 2. Or??

I think you are mixing the scenarios? The scenario we are discussing here, is when receiving separate events, i.e. it's not related to merging. So **with** the merge scenario fixed, the scenario would play out like this if we would set 1% (hardcoded, because there is not minimum dimming level):
- Initial state: off, 0%.
- Dimmer is turned on by button directly on device - at 100% brightness.
- We receive on and set brightness to 1%.
- We receive 100% and set brightness to 100%.

So in that case there would be two updates.

The other scenario is:
- Initial state: off, 0%.
- Dimmer is turned on by button directly on device - at low brightness.
- We receive on and set brightness to 1%.

With current code we would set 100% in both scenarios. And in both scenarios the brightness level as unknown, and we can interpret that in any way we see fit.

I don't feel strongly for any of the options, but I tend to slightly prefer 100% because in my most common scenario that would avoid an additional update.

The last scenario with low dimming without any dimming information is probably quite uncommon and device-specific. Did you ever see this?

jlaur replied:
> > I don't care if shown as 1% or 100%, as long as it's on,
> 
> I am Ok with the "as long as it's on" part. However I care that it should be as close to the Hue app as possible. => i.e. minimum dimming level :) .. and not 100% to just suit your scenario, and thus upset mine :(

I didn't know you had a scenario/issue - can you provide some messages documenting this, so we have more information available for making such decision? I'm under the impression that I'm really solving a corner case scenario for a device not behaving well with Hue/Zigbee standard (the hardware is Tuya-based). I wouldn't expect this to break common scenarios, but it would be better to verify that.

> > hardcoded, because there is not minimum dimming level

> But this should not be hard coded.

> I think the minimumDimmingLevel field must be available in the initial full state resource Dimming.minimumDimmingLevel in all cases -- including the example that you have observed.

As I already stated above, there is no minimum dimming level available, so it would need to be hardcoded for this scenario.

If minimum dimming level is available, do you have any cases to investigate? As otherwise we would be solving an issue we don't know if exists, and using your own wording, that would be perverse?

andrewfg replied:
I am not sure if this is what you are looking for, but the following is the case for a normal color temperature light which had been on at a high color temperature, and then turned off via a button; then the binding was disabled and re-enabled (showing the first initial state resource); then the light was turned on via the same button to a low color temperature (showing the SSE event)..

**1. Initial Full Resource**
```
[
    {
        "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
        "id_v1": "/lights/10",
        "owner": {
            "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
            "rtype": "device"
        },
        "metadata": {
            "name": "Table Lamp A",
            "archetype": "table_shade",
            "function": "functional"
        },
        "product_data": {
            "function": "functional"
        },
        "identify": {
        },
        "on": {
            "on": false
        },
        "dimming": {
            "brightness": 61.66,
            "min_dim_level": 1.0
        },
        "dimming_delta": {
        },
        "color_temperature": {
            "mirek": 153,
            "mirek_valid": true,
            "mirek_schema": {
                "mirek_minimum": 153,
                "mirek_maximum": 454
            }
        },
        "color_temperature_delta": {
        },
        "dynamics": {
            "status": "none",
            "status_values": [
                "none"
            ],
            "speed": 0.0,
            "speed_valid": false
        },
        "alert": {
            "action_values": [
                "breathe"
            ]
        },
        "signaling": {
            "signal_values": [
                "no_signal",
                "on_off"
            ]
        },
        "mode": "normal",
        "powerup": {
            "preset": "safety",
            "configured": true,
            "on": {
                "mode": "on",
                "on": {
                    "on": true
                }
            },
            "dimming": {
                "mode": "dimming",
                "dimming": {
                    "brightness": 100.0
                }
            },
            "color": {
                "mode": "color_temperature",
                "color_temperature": {
                    "mirek": 366
                }
            }
        },
        "type": "light"
    }
]
```

**2. SSE Partial Resource**
```
[
    {
        "creationtime": "2023-12-02T14:12:47Z",
        "data": [
            {
                "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
                "id_v1": "/lights/10",
                "on": {
                    "on": true
                },
                "owner": {
                    "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
                    "rtype": "device"
                },
                "type": "light"
            },
            {
                "dimming": {
                    "brightness": 35.18
                },
                "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
                "id_v1": "/lights/10",
                "owner": {
                    "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
                    "rtype": "device"
                },
                "type": "light"
            },
            {
                "color_temperature": {
                    "mirek": 454,
                    "mirek_valid": true
                },
                "id": "ba38af49-4206-47fb-b05a-b54887c12b4c",
                "id_v1": "/lights/10",
                "owner": {
                    "rid": "adeb3425-6a6b-49a0-8262-129126de7941",
                    "rtype": "device"
                },
                "type": "light"
            }
        ],
        "id": "53741602-d7e5-43b6-b5bc-527b1a8486b3",
        "type": "update"
    }
]
```

andrewfg replied:
PS and in any case look at the `Dimming.DEFAULT_MINIMUM_DIMMIMG_LEVEL` constant :)

jlaur replied:
That scenario will not be influenced by the changes, because you receive **on** and **dimming** in both messages, and at no time is there any inconsistency between them.

I have one new thought though: Instead of looking at a static state and interpreting that (on + dimming 0 = on), perhaps we should actually look at the "direction". In theory both these scenarios are possible (in practice I have only seen the first):
- Current state: off, 0.
- Message: on
- Result: on + dimming > 0 because on.

And:
- Current state: on, 100.
- Message: 0
- Result: on + dimming > 0 because on.

For the second scenario it is more likely to mean "off", because previously brightness was 100, but now it's 0. So even if we didn't receive off (yet), it seems plausible it's now off.

WDYT?

I'm wondering how to best achieve this. My first thought was to remove the current fix and test coverage and instead add a method for manipulating the received resource. So that:
- If we receive on without dimming and brightness is currently 0, we explicitly add a new Dimming object with a value > 0 (as we are still discussing).
- If we receive dimming > 0 without on and on is currently off, we explicitly add a new OnState object with value on.
- If we receive off without dimming and brightness is currently > 0, we explicitly add a new Dimming object with a value of 0.
- If we receive dimming 0 without on and on is currently on, we explicitly add a new OnState object with value off.

My only concern here is that this artificial/manipulated data would then end up in the cache, which would no longer represent the actual received state. The only way to avoid this, but still react on the "direction", would be to add the logic to a method only reading the resources and then provide this information somehow to `updateChannels()` so it could be used here. Any ideas welcome. :)

andrewfg replied:
Ok. If the Thing does not have a dimmingRange property then the bug has nothing to do with the contents, or timing, of the SSE event resources. The bug is that the Initial Full resource does not contain Dimming field.

I say 'bug' because this is indeed a real bug for this particular lamp. Fundamentally the SSE events contain a 'delta' from the initial state. But if the initial state is not fully populated then the 'delta' will always be wrong.

So we should not NOT be trying to fix the data in the SSE events, but instead we should apply a patch to fix the basis of the delta. i.e. patch the initial full state resource. So I propose that in the handler, (similar to the `applyDeviceSpecificWorkArounds()`), you should add a method that does the following to fix the initial full state resource.

```
if (! hasSparseData && dimming == null) {
  dimming = new Dimming().setMinimumDimmingLevel(Dimming.DEFAULT_MINIMUM_DIMMIMG_LEVEL);
}
```

jlaur replied:
I have no problem with that, it seems like a good idea and this would then obsolete this logic:

https://github.com/openhab/openhab-addons/blob/f0c8defabb6d7237a8ebb72686428b6ac8645ff3/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2ThingHandler.java#L402

since we would then always have a dimming level present, either actual or defaulted.

However, this is only a side-topic.

andrewfg replied:
> we would then always have a dimming level present, either actual or defaulted.

Indeed.

.. and potentially the same would apply to gamut and mirek-schema. However, this is **also** only a side-topic. :)

jlaur replied:
Hmm, actually I think the defaulting logic of minimum dimming should be "on the fly" rather than cached. Otherwise it could cause problem with not overwriting it in cache later if it "appears late". And it would also cause it to be shown as a property, which I don't think it should when it is actually not available. It should be visible in properties that there is no actual minimum dimming declared.

jlaur replied:
Maybe let's back to the main topic - just in case you missed this: https://github.com/openhab/openhab-addons/pull/15905#discussion_r1412816234 :)

andrewfg replied:
> defaulting logic of minimum dimming should be "on the fly" rather than cached

I don't recall the full details, but the bridge runs a scheduled task `updateSelf()` periodically to do a full state update on all things. So this 'patch' would be called in there somewhere..

jlaur replied:
Just to be clear, that's a side-topic and will not fix the issue fixed by 52ea4f2e7bd224992c8d883c17223124251a50f9. The problem is that the SSE event contains only **on**, but not **dimming**. Dimming will either arrive in a delayed event, or not at all. So at the moment when **on** is received and current cached brightness (either from initial state or last SEE update) is 0, we have inconsistency.

andrewfg replied:
> `brightness = current cached brightness != 0 ? current cached brightness : minimum dimming level;`

But do this conversion in the thing handler (where you can access the cache) rather than where it is currently located in the resource.

jlaur replied:
Yes, but we are back to:
- https://github.com/openhab/openhab-addons/pull/15905#discussion_r1412816234 (to not only consider current state, but also its context/"direction")
- https://github.com/openhab/openhab-addons/pull/15905#discussion_r1412792663 (minimum dimming vs. 100% - while remembering this is a specific scenario unlike your "normal" scenario, which will not be affected)

andrewfg replied:
The 'direction' and/or the choice 'minimum dimming vs. 100%' are both/either only wild guesses, and not based on any facts or received data. Whereas cached value is definitely a fact. And if no later event comes then cached value remains a fact. If cached value is 0 (yet state is on) the Hue app uses minimum dimming value. => So follow the same logic.

andrewfg replied:
> If we receive on without dimming and brightness is currently 0

Use minimum dimming

> If we receive dimming > 0 without on and on is currently off

It is off, until proven otherwise

> If we receive off without dimming and brightness is currently > 0

It is off

> If we receive dimming 0 without on and on is currently on

Does this ever happen? I doubt it..

jlaur replied:
> > If we receive dimming > 0 without on and on is currently off
> 
> It is off, until proven otherwise

By definition in openHAB, if brightness > 0, it is **on**. So if the newly received value is **dimming** > 0, then it should be **on**, no matter what the previous **on** value was. It would be inconsistent to disregard this information.

andrewfg replied:
> By definition in openHAB, if brightness > 0, it is on

I think you are getting tired, perhaps take a break to think carefully about this.
- Your above statement does indeed apply when OH sends an HSBType command with B > 0. This should certainly send an On command to the light. (And it does).
- It does NOT apply the other way round. If the bridge sends Dimming.brightness 0 to the binding, the lamp is still on -- albeit at very low brightness i.e. minimum level -- but it is still On!
- You shall not mis-interpret Dimming.brightness 0 events as an Off state report. You shall wait until the bridge specifically sends an Off state.
- If you a receive Dimming.brightness 0 event and have not (yet) received an Off, it is ok to mess with the B part of the returned HSBType to satisfy OH (set B to min dim level). But is is not Ok to change Dimming.brightness to anything other than that which was received.
- If you receive an Off event (and have or have not received a Dimming.brightness 0 value), then it is ok to set the B part of the returned HSBType to 0 to satisfy OH. But is is not Ok to change Dimming.brightness to anything other than that which was received.

jlaur replied:
Since there was some misunderstanding [here](https://github.com/openhab/openhab-addons/issues/15700#issuecomment-1835676939) and we do not have consensus or an implementation, I have removed disputable commit from this PR to not prolong the review process as I think the part we do agree upon wrt the merge algorithm is more or less in a ready state.

I have extracted the remaining issue to #15995.

andrewfg commented:
@jlaur I am a bit confused. You seem to have deleted some of your comments and my responses. (Probably a good thing). But can you please just update me on where you stand with the scope of this PR? I assume that you want to just handle the merge of multiple SSE events in this PR? And handle the other topic elsewhere? Is that correct? In which case I will do a final review of the PR on that basis. Is it all ready for my final review.

And concerning the other topic, how do you want to proceed? As before I am willing to propose some code if you like.

jlaur commented:
> @jlaur I am a bit confused. You seem to have deleted some of your comments and my responses. (Probably a good thing).

Not intentionally at least, but I resolved the comment we used for discussing today, since that is now a separate issue.

> But can you please just update me on where you stand with the scope of this PR? I assume that you want to just handle the merge of multiple SSE events in this PR?

Yes.

> And handle the other topic elsewhere? Is that correct? In which case I will do a final review of the PR on that basis. Is it all ready for my final review.

Yes, see above.

> And concerning the other topic, how do you want to proceed? As before I am willing to propose some code if you like.

You are welcome to propose.

andrewfg reviewed (approved):
LGTM

andrewfg commented:
@jlaur I will create the PR tomorrow..

jlaur commented:
@lolodomo - we have three different Hue bugfixes at the moment, all concerning some quite related symptoms with undesired state updates. #15984 is blocked by a dependency towards a bugfix in core. #15999 needs test and review, so also not ready. However, this one is ready. Any chance of a quick review? 😉 That would help testing and managing the combined fix which is really needed, and therefore would be much appreciated. Most of the lines are test coverage.

jlaur commented:
Thanks for the sanity check and merge, @kaikreuzer. 🙂
